lib/webauthn/attestation_statement/android_safetynet.rb

```diff
@@ -1,32 +1,36 @@
 # frozen_string_literal: true
 
-require "android_safetynet/attestation_response"
+require "safety_net_attestation"
 require "openssl"
 require "webauthn/attestation_statement/base"
 
 module WebAuthn
   module AttestationStatement
     # Implements https://www.w3.org/TR/webauthn-1/#sctn-android-safetynet-attestation
     class AndroidSafetynet < Base
-      def valid?(authenticator_data, client_data_hash)
-        valid_response?(authenticator_data, client_data_hash) &&
+      def valid?(authenticator_data, client_data_hash, time)
```

Comment:
What was the issue that made this `time` argument necessary?

Reply:
It replaces libfaketime for the tests in `authenticator_attestation_response_spec.rb` by injecting the time dependency used for verification instead. `OpenSSL::X509::{Store,StoreContext}` has OpenSSL  get the time from system directly unless specified with `#time=` for verifying certificates, which lead to the libfaketime workaround. In case of SafetyNet the time also needs to be set correctly for verification of the `timestampMs`.

There's multiple roads that lead to Rome but in https://github.com/bdewater/safety_net_attestation/commit/2aa549f97e5f744506dcf9aec707183cee435d03 I implemented injectable Time objects as part of the safety_net_attestation API. It's IMO cleaner than using RSpec's `and_wrap_original` which was already much better than `libfaketime`. But if you follow the advice of 'storing the returned trust path' as part of the registration ceremony, this allows you to re-verify the attestation statement again at some future date by using the `created_at` timestamp.

For SafetyNet it is not very interesting to re-verify attestation, but for example Android Key attestation or statements from the Fido MDS have mechanisms for revocation. This means having an interface that lends itself to easily injecting the verification timestamp allows you to warn users their authenticator has been compromised. This would be the WebAuthn alternative to incorporating https://haveibeenpwned.com/API/ in your app.

Reply:
Gotcha.

Thanks for the detailed explanation.

Pretty cool that we can now replace long-time libfaketime hack with something else! :-)

I think I would prefer to go with the `and_wrap_original` rspec mock instead of making a change to all attestation statement's `#valid?` method signature at this point. I understand the point that maybe allowing for custom time can be useful for some use cases in the future, but not sure we would want that right now and as part of a "refactor" PR.

As much as we can keep this PR being a "refactor" only, the better :-)

Reply:
Alright. I'm planning on adding certificate verification to Android Key attestation soon (right now there's nothing chaining it up to a root of trust) so will circle back in the near future.